### ISSUE TITLE
DEDataArrays and saveat don't work together

### ISSUE THREAD

**Issue by BarrOff** [JuliaDiffEq/StochasticDiffEq.jl#247]
Hello everyone,

playing around with DEDataArrays I encountered an error when combining it with the <saveat> solver option. @ChrisRackauckas said I should open an issue, so here it is.

This works:
`sol = solve(prob,callback = cbs,tstops=tstop,EM(),dt=dt)`
whereas
`sol = solve(prob,callback = cbs,tstops=tstop,EM(),dt=dt,saveat=collect(8:0.1:10))`
creates an error:
`ERROR: LoadError: MethodError: Cannot `convert` an object of type Array{Float64,1} to an object of type SimType{Float64}`

A file containing a MWE is attached. I took the example from the [DiffEq-Specific Array Types](https://docs.juliadiffeq.org/stable/features/diffeq_arrays/) page and converted it to an SDE.

I don't know Julia that well but I'd like to help as much as I can by testing fixes or similar things.
PS: thanks for your great work JuliaDiffEq-team! :-)

[error_example.txt](https://github.com/JuliaDiffEq/StochasticDiffEq.jl/files/3896338/error_example.txt)

**Comment by ChrisRackauckas** [comment]
Thanks for the issue! I have a lot on my plate so this is not a top priority right now, so I do not plan to get to it this week. However, feel free to ping me in a few weeks and I'll probably find a fix. Sorry that's the case but I hope being honest helps you plan better. Cheers!

**Comment by BarrOff** [comment]
@ChrisRackauckas Would it be possible to look into it till the end of next week? I really need this to work for my current project

**Comment by devmotion** [comment]
I quickly ran your example and reproduced your error. The problem is the interpolation that is necessary if you specify `saveat` - the integration algorithm takes certain steps but you want to save the solution at some other time points. More precisely, only if you specify `saveat`

https://github.com/JuliaDiffEq/StochasticDiffEq.jl/blob/d0cb51e7426374b09b8e630797a7970f4685293c/src/integrators/integrator_utils.jl#L74

is executed if you save your solution. The problem now is that the interpolation returns a regular vector instead of an instance of `SimType` (which is the type of your states), which then can't be added to your vector of solution states of type `SimType` since there exists no method to convert a regular vector to an instance of `SimType` (that explains the error message you get).

I'm not sure what's the proper way to solve this issue. To me it feels reasonable that the interpolated value is actually just a regular vector - I mean, how could one know how to interpolate arbitrary other fields in your data array (and it might not even make sense to perform any interpolation on them)? On the other hand, of course, since your state is a data array it's not completely unreasonable to create a solution that saves instances of your custom data array type - maybe there's some interesting information contained in your other fields that you would like to keep?

So basically I see two possible approaches:
- Always just save `u.x` in the case of data arrays and discard any other information. That should give consistent solutions, regardless whether you specify `saveat` or not, but would of course only save regular arrays and discard all other information in your data array
- Demand from the user to define a `Base.convert(::Type{SimType{T}}, ::Vector{T}) where T` method if `saveat` is specified. E.g., you could solve your issues by defining
```julia
Base.convert(::Type{SimType{T}}, x::Vector{T}) where T = SimType(x, zero(T))
```
(no idea if that makes sense :smile:).

**Comment by ChrisRackauckas** [comment]
DEDataArrays have the assumption that the other variables are discrete, so they should be constant in an interpolation.  In that interpretation, taking the value "from the left" makes sense.

**Comment by ChrisRackauckas** [comment]
MWE:

```julia
u0 = SimType([10.0;10.0], 0.0)
u1 = SimType([10.0;10.0], 0.0)
Θ = 0.5
@. (1-Θ)*u0 + Θ*u1
```

**Comment by ChrisRackauckas** [comment]
Boomdiggity.

```julia
const DEDataArrayStyle = Broadcast.ArrayStyle{DEDataArray}
Base.BroadcastStyle(::Type{<:DEDataArray}) = Broadcast.ArrayStyle{DEDataArray}()
Base.BroadcastStyle(::Broadcast.ArrayStyle{DEDataArray},::Broadcast.ArrayStyle) = Broadcast.ArrayStyle{DEDataArray}()
Base.BroadcastStyle(::Broadcast.ArrayStyle,::Broadcast.ArrayStyle{DEDataArray}) = Broadcast.ArrayStyle{DEDataArray}()
Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{DEDataArray}},::Type{ElType}) where ElType = similar(find_dedata(bc))

find_dedata(bc::Base.Broadcast.Broadcasted) = find_dedata(bc.args)
function find_dedata(args::Tuple)
  !isempty(args) && find_dedata(find_dedata(args[1]), Base.tail(args))
end
find_dedata(x) = x
find_dedata(a::DEDataArray, rest) = a
find_dedata(::Any, rest) = find_dedata(rest)

u0 = SimType([10.0;10.0], 0.0)
u1 = SimType([15.0;20.0], 0.0)
Θ = 0.5
@. (1-Θ)*u0 + Θ*u1
```

Next DiffEqBase release will close probably all of the DEDataArrays issues with this.

**Comment by devmotion** [comment]
> DEDataArrays have the assumption that the other variables are discrete, so they should be constant in an interpolation. In that interpretation, taking the value "from the left" makes sense.

Isn't that interpretation wrong if one integrates backwards?

**Comment by BarrOff** [comment]
@devmotion the second approach (defining a convert function) looses the concrete extra variables as I see no way of retrieving their current values. So you would have to give them generic values which contradicts the point of using DEDataArays.

**Comment by devmotion** [comment]
But at least the solution above (haven't checked the PR to DiffEqBase yet) is even worse, since it randomly keeps some values. It just depends on the order of the operation, there's no way to enforce that we take the value "from the left" IMO:
```julia
u0 = SimType([10.0;10.0], 1.0)
u1 = SimType([15.0;20.0], 2.0)
Θ = 0.5

(@. (1-Θ)*u0 + Θ*u1).f1 # 1.0
(@. Θ*u1 + (1-Θ)*u0).f1 # 2.0
```

**Comment by ChrisRackauckas** [comment]
yeah, there's no way to enforce it. My plan is to get this done because it works for most people though... and then open an issue for making that more controllable for the future, but at a lower priority.

**Comment by ChrisRackauckas** [comment]
See https://github.com/JuliaDiffEq/DiffEqBase.jl/blob/master/test/downstream/data_array_regression_tests.jl for a corrected example.
